Add support for untyped metrics from Ops Agent

damemi opened the pull request:
This adds functionality to automatically parse the attribute set by the Ops Agent's prometheus receiver to indicate untyped prometheus metrics (https://github.com/GoogleCloudPlatform/opentelemetry-operations-collector/pull/170). When this attribute is found in a Gauge metric, the time series will be double-exported as a Gauge and a Cumulative. The special attribute will be dropped from metric labels.

This behavior is ~~on by default~~ behind a featuregate `gcp.untyped_double_export` and is only meant for internal use until an upstream approach to untyped metrics is implemented.

dashpole commented:
> This behavior is on by default and is only meant for internal use until an upstream approach to untyped metrics is implemented.

Can we put this behind an alpha feature gate?  Otherwise, I would considering it a breaking change when we remove the behavior

damemi commented:
> > This behavior is on by default and is only meant for internal use until an upstream approach to untyped metrics is implemented.
> 
> Can we put this behind an alpha feature gate? Otherwise, I would considering it a breaking change when we remove the behavior

We can, but the mechanism (the attribute) isn't user-facing, so I think changing it wouldn't be breaking. Of course there's nothing to stop users from adding this attribute themselves anyway and hacking it.

dashpole commented:
Should this be specific to the GMP Exporter?  I'm OK with this version being in the googlecloud exporter, but I think we want it in GMP parts eventually.

damemi commented on exporter/collector/metrics.go:
```diff
@@ -1227,7 +1240,38 @@ func (m *metricMapper) gaugePointToTimeSeries(
+					Value: value,
+				}},
+				Metric: &metricpb.Metric{
+					Type: t,
```
Good catch, it looks like GMP uses a special [`unknown`](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/107933594665ab14ac69f21a66fc6f2c71f130a7/pkg/export/series_cache.go#L326) suffix [when it duplicates the metric](https://github.com/GoogleCloudPlatform/prometheus-engine/blob/cfbb931e82bbefe852a42e8d2eadc5fcf207b38b/pkg/export/series_cache.go#L421-L429). Maybe we should change our metric name function to check for this label

dashpole replied:
Note that it has different suffixes for the gauge portion of the unknown metric (`unknown` and the counter portion of the unknown metric (`unknown:counter`).

damemi replied:
I think I'm going to refactor this, I think a better approach would be to copy the untyped Gauge pmetric entry into an extra Sum metric, and let the existing path after that handle everything. Then `GetMetricName()` can handle both.

I think using an actual Sum from pdata will also make it easier to handle @ridwanmsharif's question https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/668#discussion_r1261612291. @dashpole wdyt?

dashpole replied:
Works for me, but I'm not convinced the logic should live in the base exporter.  If someone used the prometheus receiver with our base exporter, they would get duplicate timeseries errors.  We need the GMP logic of adding suffixes to prevent duplicate timeseries problems.

damemi replied:
We can at least move the gate to the GMP exporter so the logic is only active there. Not sure if we could move the logic itself. Maybe through the "extra metrics" approach we did for target and scope info metrics actually

damemi replied:
I was able to move everything using the extra metrics approach. Ptal, thanks!

ridwanmsharif commented on exporter/collector/metrics.go:
```diff
@@ -1227,7 +1240,38 @@ func (m *metricMapper) gaugePointToTimeSeries(
+				ValueType:  valueType,
+				Points: []*monitoringpb.Point{{
+					Interval: &monitoringpb.TimeInterval{
+						EndTime: timestamppb.New(point.Timestamp().AsTime()),
```
Just wanted to confirm this, but not setting the start time will make this normalize the same way GMP does with these counters? 

That is, when the value goes down, treat it as a reset?

damemi replied:
I changed this, but it's still doing basically the same thing in the new code so to answer your question: yes it should. This is the normalization function: https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/a19d9eab7eac8a63dcf2bf443b481b48ac62570e/exporter/collector/internal/normalization/standard_normalizer.go#L261 (called [from here on a Sum](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/a19d9eab7eac8a63dcf2bf443b481b48ac62570e/exporter/collector/metrics.go#L1158), which the updates I just pushed use for the Cumulative instead of a Gauge)

You can see this in the fixtures, where I had to add [2 new untyped input data points](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/4a74e8973dfd0ec128deba4923c077ee90b30127/exporter/collector/integrationtest/testdata/fixtures/metrics/google_managed_prometheus_expect.json#L96-L162) just to get [one output unknown cumulative](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/4a74e8973dfd0ec128deba4923c077ee90b30127/exporter/collector/integrationtest/testdata/fixtures/metrics/google_managed_prometheus_untyped_expect.json#L530-L562). This is because the first data point gets normalized as a reset point and dropped by our exporter.

As a side note, I had to update our fixture generators to test this because they currently normalize all timestamps to be rightNow() (in order to work with testing against the actual GCM api). But I needed to bypass that for this test to get the cumulative normalization to show up.

damemi commented:
Collector tests are failing from something in the "skip timestamp update" hack I put in... maybe I can't do that

damemi commented on exporter/collector/metrics.go:
```diff
@@ -1300,7 +1344,9 @@ func (me *metricMapper) convertToBoolIfMetricKindSupported(
 	ls := make(labels, attrs.Len())
 	attrs.Range(func(k string, v pcommon.Value) bool {
-		ls[sanitizeKey(k)] = sanitizeUTF8(v.AsString())
+		if k != GCPOpsAgentUntypedMetricKey {
```
updated, dropping this attribute is now handled by the GMP custom naming function (and behind the feature gate)

damemi commented on exporter/collector/integrationtest/testdata/fixtures/metrics/google_managed_prometheus_untyped_expect.json:
```diff
@@ -0,0 +1,980 @@
+        },
+        {
+          "metric": {
+            "type": "prometheus.googleapis.com/fake_untyped_metric_total/unknown:counter",
```
here is the expected counter... of note:

* `_total` is appended (I think this comes from the upstream prom normalization that's currently on be default)
* 2 data points for `fake_untyped_metric` as a gauge, because there are 2 input points (first Sum is dropped for normalization)

dashpole replied:
> I think this comes from the upstream prom normalization that's currently on be default

Should we update our collector dependency?  I don't think that is supposed to be the case anymore.

dashpole replied:
Or we can leave it how it is, and update later

damemi replied:
I did a test bump on this branch and re-ran the fixture and the `_total`/units suffixes are dropped in the new version. I'll open a separate bump PR after I merge this

damemi commented on exporter/collector/integrationtest/testdata/fixtures/metrics/untyped_prometheus_metrics_expect.json:
```diff
@@ -0,0 +1,1234 @@
+            "type": "workload.googleapis.com/fake_untyped_metric",
+            "labels": {
+              "ex_com_lemons_untyped": "13",
+              "prometheus_untyped_metric": "true",
```
this is the fixture for testing the featuregate. this label would be dropped in the gmp exporter if the gate was enabled

damemi commented:
Final update now that all tests are green: using 2 gauge entries to get the normalized cumulative didn't work with the actual GCM integration test, because the 2nd gauge caused a duplicate timeseries error ([see failure here](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/668/checks?check_run_id=15050400101)). Added https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/668/commits/78303434923a6d0785f2d7f63d4769dff4566f46 to disable cumulative normalization, use 1 gauge entry, and remove the "skip timestamp update" option I added earlier.

However the lack of startTimestamp on the input gauge caused it to be initialized to the 0 time [by this line](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/blob/ff27f899ee4b771e7a7ac90e517841f07465b2b5/exporter/collector/metrics.go#L1164), which caused an invalid timeseries error from GCM ([see that failure here](https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/668/checks?check_run_id=15054768170)). To fix that I pushed https://github.com/GoogleCloudPlatform/opentelemetry-operations-go/pull/668/commits/505a1cfb61500b8dd4395ba20d0f6e8bd69b8a0d to add a startTimestamp to the input gauge and more importantly copy that startTimestamp to the new point. I think this is valid for use cases where untyped metrics might want to manually register a reset point.

Now that this is green, I'm merging. Will open another issue to check on our other cumulative tests to make sure we're not dropping points in those.
